services/datamanager/builtin/sync_test.go

```diff
@@ -0,0 +1,652 @@
+	datasync.RetryExponentialFactor = 1
+	datasync.InitialWaitTimeMillis.Store(int32(20))
+	captureTime := time.Millisecond * 300
+	syncTime := time.Millisecond * 100
+
+	tests := []struct {
+		name                  string
+		dataType              v1.DataType
+		manualSync            bool
+		scheduledSyncDisabled bool
+		serviceFailAt         int
+		numFails              int
+	}{
+		{
+			name:     "Previously captured tabular data should be synced at start up.",
+			dataType: v1.DataType_DATA_TYPE_TABULAR_SENSOR,
+		},
+		{
+			name:     "Previously captured binary data should be synced at start up.",
+			dataType: v1.DataType_DATA_TYPE_BINARY_SENSOR,
+		},
+		{
+			name:                  "Manual sync should successfully sync captured tabular data.",
+			dataType:              v1.DataType_DATA_TYPE_TABULAR_SENSOR,
+			manualSync:            true,
+			scheduledSyncDisabled: true,
+		},
+		{
+			name:                  "Manual sync should successfully sync captured binary data.",
+			dataType:              v1.DataType_DATA_TYPE_BINARY_SENSOR,
+			manualSync:            true,
+			scheduledSyncDisabled: true,
+		},
+		{
+			name:       "Running manual and scheduled sync concurrently should not cause data races or duplicate uploads.",
+			dataType:   v1.DataType_DATA_TYPE_TABULAR_SENSOR,
+			manualSync: true,
+		},
+		{
+			name:          "If tabular sync fails part way through, it should be resumed without duplicate uploads",
+			dataType:      v1.DataType_DATA_TYPE_TABULAR_SENSOR,
+			serviceFailAt: 2,
+		},
+		{
+			name:          "If binary sync fails part way through, it should be resumed without duplicate uploads",
+			dataType:      v1.DataType_DATA_TYPE_BINARY_SENSOR,
+			serviceFailAt: 2,
+		},
+		{
+			name:     "If tabular uploads fail transiently, they should be retried until they succeed.",
+			dataType: v1.DataType_DATA_TYPE_TABULAR_SENSOR,
+			numFails: 2,
+		},
+		{
+			name:     "If binary uploads fail transiently, they should be retried until they succeed.",
+			dataType: v1.DataType_DATA_TYPE_BINARY_SENSOR,
+			numFails: 2,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Set up server.
+			tmpDir, err := os.MkdirTemp("", "")
+			test.That(t, err, test.ShouldBeNil)
+			defer func() {
+				err := os.RemoveAll(tmpDir)
+				test.That(t, err, test.ShouldBeNil)
+			}()
+			rpcServer, mockService := buildAndStartLocalSyncServer(t, tc.serviceFailAt, tc.numFails)
+			defer func() {
+				err := rpcServer.Stop()
+				test.That(t, err, test.ShouldBeNil)
+			}()
+
+			// Set up data manager.
+			dmsvc := newTestDataManager(t)
+			dmsvc.SetSyncerConstructor(getTestSyncerConstructor(rpcServer))
+			var cfg *config.Config
+			if tc.dataType == v1.DataType_DATA_TYPE_TABULAR_SENSOR {
+				cfg = setupConfig(t, enabledTabularCollectorConfigPath)
+			} else {
+				cfg = setupConfig(t, enabledBinaryCollectorConfigPath)
+			}
+
+			// Set up service config.
```

Comment:
[nit] specify that we're only capturing here

Reply:
Done